src/matrix-to.js

```diff
@@ -14,21 +14,107 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
+import MatrixClientPeg from "./MatrixClientPeg";
+
 export const host = "matrix.to";
 export const baseUrl = `https://${host}`;
 
+// The maximum number of servers to pick when working out which servers
+// to add to permalinks. The servers are appended as ?via=example.org
+const MAX_SERVER_CANDIDATES = 3;
+
 export function makeEventPermalink(roomId, eventId) {
-    return `${baseUrl}/#/${roomId}/${eventId}`;
+    const serverCandidates = pickServerCandidates(roomId);
+    return `${baseUrl}/#/${roomId}/${eventId}?${encodeServerCandidates(serverCandidates)}`;
 }
 
 export function makeUserPermalink(userId) {
     return `${baseUrl}/#/${userId}`;
 }
 
 export function makeRoomPermalink(roomId) {
-    return `${baseUrl}/#/${roomId}`;
+    const serverCandidates = pickServerCandidates(roomId);
+    return `${baseUrl}/#/${roomId}?${encodeServerCandidates(serverCandidates)}`;
 }
 
 export function makeGroupPermalink(groupId) {
     return `${baseUrl}/#/${groupId}`;
 }
+
+export function encodeServerCandidates(candidates) {
+    if (!candidates) return '';
+    return `via=${candidates.map(c => encodeURIComponent(c)).join("&via=")}`;
+}
+
+export function pickServerCandidates(roomId) {
+    const client = MatrixClientPeg.get();
+    const room = client.getRoom(roomId);
+    if (!room) return [];
+
+    // Permalinks can have servers appended to them so that the user
+    // receiving them can have a fighting chance at joining the room.
+    // These servers are called "candidates" at this point because
+    // it is unclear whether they are going to be useful to actually
+    // join in the future.
+    //
+    // We pick 3 servers based on the following criteria:
+    //
+    //   Server 1: The highest power level user in the room, provided
+    //   they are at least PL 50. We don't calculate "what is a moderator"
+    //   here because it is less relevant for the vast majority of rooms.
+    //   We also want to ensure that we get an admin or high-ranking mod
+    //   as they are less likely to leave the room. If no user happens
+    //   to meet this criteria, we'll pick the most popular server in the
+    //   room.
+    //
+    //   Server 2: The next most popular server in the room (in user
+    //   distribution). This will probably be matrix.org in most cases
```

Comment:
I'd avoid mentioning matrix.org here; it's very much a misfeature that public rooms are dominated by matrix.org users today, plus this ignores usages in private federations.